pkg/kubefed2/federate/federate.go

```diff
@@ -190,44 +197,50 @@ func getTargetResource(hostConfig *rest.Config, typeConfig *fedv1a1.FederatedTyp
 	return resource, nil
 }
 
-func createFedResource(hostConfig *rest.Config, typeConfig *fedv1a1.FederatedTypeConfig, template *unstructured.Unstructured, dryrun bool) (*unstructured.Unstructured, error) {
+func FederatedResourceFromTargetResource(typeConfig typeconfig.Interface, template *unstructured.Unstructured) (*unstructured.Unstructured, error) {
 	fedAPIResource := typeConfig.GetFederatedType()
-	fedClient, err := ctlutil.NewResourceClient(hostConfig, &fedAPIResource)
-	if err != nil {
-		return nil, errors.Wrapf(err, "Error creating client for %s", fedAPIResource.Kind)
-	}
 
-	targetKind := typeConfig.GetTarget().Kind
-	if targetKind == ctlutil.ServiceAccountKind {
+	if typeConfig.GetTarget().Kind == ctlutil.ServiceAccountKind {
 		unstructured.RemoveNestedField(template.Object, ctlutil.SecretsField)
 	}
 
 	qualifiedName := ctlutil.NewQualifiedName(template)
-	resourceNamespace := ""
-	if typeConfig.GetTarget().Kind == ctlutil.NamespaceKind {
-		resourceNamespace = qualifiedName.Name
-	} else {
-		resourceNamespace = qualifiedName.Namespace
-	}
+	resourceNamespace := getNamespace(typeConfig, qualifiedName)
 	fedResource := &unstructured.Unstructured{}
-	SetBasicMetaFields(fedResource, fedAPIResource, qualifiedName.Name, resourceNamespace, "")
+	SetBasicMetaFields(fedResource, fedAPIResource, qualifiedName.Name, resourceNamespace, template.GetGenerateName())
```

Comment:
This was needed by test, and I thought it might be ok as the actual resource which we are federating can also have a `generateName`. On second thoughts I have removed it from here and setting it explicitly in tests.